### PR TITLE
add isStrictlyMonotonic for ranges

### PR DESCRIPTION
**Pull request by wilzbach**
Re-submission of the reverted #4025. Addressing @andralex comments

> "Strictly sorted" is a confusing misnomer. It's a misnomer because a given set of items may not be "strictly" or "non-strictly" sorted. Sortedness and having no duplicates are distinct properties. It's confusing because "strict" in sorting usually refers to the comparison predicate.

It was named `strict` because there is `strictlyOrdered` - you might want to rename that too then?
-> Renamed to `isStrictlyMonotonic`

> The implementation is wasteful. This is a one-liner:
> return adjacentFind!((a, b) => !binaryFun!less(a, b)).empty;

1) it's `findAdjacent`
2) Couldn't `isSorted` use it too with `<=` then?

Wow that's pretty clever! Changed.
 
> I'm not sure there's enough necessity for this functionality to make it a separate function.

Well there is `strictlyOrdered` too!

The other change to this PR is that now the new ticks "`code`" are used instead of `$(D code)`.

### CODE REVIEW

**Comment by JackStouffer** [comment]
Should have a test to verify `@nogc` inference. For example:

```
@nogc @safe nothrow pure unittest
{
    static immutable a = [1, 2, 3];
    assert(a.isStrictlyMonotonic);
}
```

**Comment by greenify** [comment]
> Should have a test to verify @nogc inference. For example:

Okay - then `isSorted` should have the same, right?
I saw that not many methods have these @nogc inference tests - is there any value to go through phobos and add such tests?

**Comment by JakobOvrum** [comment]
`@nogc` is a relatively new attribute. It's possible that older generic code was automatically inferred to be `@nogc` when `@nogc` was introduced despite not being explicitly designed for it. Adding a test formalizes support for the attribute and promises that future changes to the implementation won't break the attribute guarantee for those instantiations or that use-pattern (although we don't have a standard way to reflect this in the documentation).

Adding `@nogc` unit tests opportunistically when working on older code makes sense to me unless we want to keep the door open to using GC in the implementation.

**Comment by greenify** [comment]
I forgot to add this method to the package booktable of `std.algortihm` -> Updated.

@andralex do you think adding this function adds enough value to phobos?
A pro point is that there are `ordered` and `strictlyOrdered`.

**Comment by andralex** [comment]
OK let's put this in. I approve the addition. One idea to help with documentation bloating - could you please group `isSorted` together with `isStrictlyMonotonic`? Then they share most documentation and the only change needed is "Unlike `isSorted`, `isStrictlyMonotonic` does not allow for equal values, i.e. values for which both `less(a, b)` and `less(b, a)` are false."

**Comment by greenify** [comment]
> could you please group isSorted together with isStrictlyMonotonic?

done.

**Comment by greenify** [comment]
So we go with this function or directly with #4134?

**Comment by wilzbach** [comment]
Ping - any other objections? I would suggest we use this and then continue the discussion whether it can be combined with ordered at #4134?

File: std/algorithm/sorting.d [hunk 1]
```diff
@@ -205,13 +224,63 @@ bool isSorted(alias less = "a < b", Range)(Range r) if (isForwardRange!(Range))
...
+{
+    static immutable a = [1, 2, 3];
+    assert(a.isSorted);
+}
+
+/// ditto
+bool isStrictlyMonotonic(alias less = "a < b", Range)(Range r)
+if (isForwardRange!Range)
+{
+    import std.algorithm.searching;
```
Comment (JackStouffer) [inline std/algorithm/sorting.d]: Should be a selective import

**Comment by wilzbach** [comment]
All comments addressed.

**Comment by JackStouffer** [comment]
LGTM

**Comment by wilzbach** [comment]
(added entry to the changelog)

**Comment by wilzbach** [comment]
ping pong for objections

**Comment by wilzbach** [comment]
> OK let's put this in. I approve the addition.

Just rebased (changelog got updated) - anyone else objecting?
Randomly ping @MetaLang @Hackerpilot @DmitryOlshansky @braddr

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by wilzbach** [comment]
> LGTM

I just rebased as the changelog updated itself again.
On a related matter: it would be great to have a changelog folder, s.t. we don't get merge conflicts and people start to put their changelog entry directly in the PR.

**Comment by Hackerpilot** [comment]
Auto-merge toggled on

**Comment by wilzbach** [comment]
Thanks!

**Comment by JackStouffer** [comment]
> On a related matter: it would be great to have a changelog folder, s.t. we don't get merge conflicts and people start to put their changelog entry directly in the PR.

Honestly, it should just be on the wiki.

**Comment by wilzbach** [comment]
> Honestly, it should just be on the wiki.

I just posted to the [newsgroup](http://forum.dlang.org/post/pbfgjdvhvfgobcpnlvfb@forum.dlang.org), s.t. we don't hijack this issue ;-)

**Pull request merged**